src/algo.rs

```diff
@@ -0,0 +1,301 @@
+    fn dominates(&self, other: &Self) -> bool {
+        self.arr_time <= other.arr_time && self.arr_time >= other.dep_time
+    }
+
+    fn is_non_dominated(&self, other_opt: Option<&Self>) -> bool {
+        // By construction, we know self.dep_time >= other.dep_time
+        match other_opt {
+            Some(other) => self.arr_time <= other.arr_time,
+            None => true,
+        }
+    }
+}
+
+fn arrival_time_with_stop_change(profiles: &[Profile], c: &Connection) -> Option<u16> {
+    let transfer_duration = 5;
+    profiles
+        .iter()
+        .take_while(|p| p.dep_time > c.arr_time + transfer_duration)
+        .last()
+        .map(|p| {
+            if p.out_connection.is_some() {
+                p.arr_time
+            } else {
+                // If this is the very last connection to target, it gives us the arrival time
+                c.arr_time
+            }
+        })
+}
+
+trait Incorporate {
+    fn incorporate(&mut self, candidate: Profile) -> bool;
+}
+
+impl Incorporate for Vec<Profile> {
+    fn incorporate(&mut self, candidate: Profile) -> bool {
+        // The profiles are ordered in decreasing dep_time
+        // We want to insert à the right position, if not dominated
+        // and delete dominated solutions
+        let mut earlier_profiles = Vec::new();
+
+        // We first remove all the profiles earlier
+        while let Some(last) = self.pop() {
+            if last.dep_time >= candidate.dep_time {
+                self.push(last);
+                break;
+            } else if !candidate.dominates(&last) {
+                earlier_profiles.push(last);
+            }
+        }
+
+        let incorporated = candidate.is_non_dominated(self.last());
+        if incorporated {
+            self.push(candidate);
+        }
+        // Let’s put back the previous
+        for earlier_profile in earlier_profiles.into_iter().rev() {
+            self.push(earlier_profile);
+        }
+
+        incorporated
+    }
+}
+
+// It returns all the possible routes, from all possible nodes to the given destination
+pub fn compute(timetable: &Timetable, destination: usize) -> Vec<Vec<Profile>> {
+    let mut arr_time_with_trip: Vec<_> = timetable.trips.iter().map(|_| None).collect();
+    let mut profiles: Vec<_> = timetable.stops.iter().map(|_| Vec::new()).collect();
+    profiles[destination].push(Default::default());
+
+    for (conn_index, c) in timetable.connections.iter().enumerate() {
+        // Case 1: walking to target, TODO
+
+        // Case 2: Staying seated in the trip, we will reach the target at `t2`
+        let t2 = arr_time_with_trip[c.trip];
+
+        // Case 3: Transfering in the same stop, we look up the earliest compatible arrival
+        let t3 = arrival_time_with_stop_change(&profiles[c.arr_stop], c);
+
+        if let Some(t) = [t2, t3].iter().filter_map(|t| *t).min_by(|a, b| a.cmp(b)) {
+            let candidate = Profile {
+                out_connection: Some(conn_index),
```

Comment:
what do you use out_connection for ? it seems to be only used as a boolean (I don't like the untyped index)

Reply:
It will be used later to reconstruct the path, for now it’s indeed no really needed yet